### PR TITLE
Add elexonpy to requirements and display forecasts for different process types

### LINKED ISSUE #172: Add elexon forecast to graph
## Detailed Description
Add elexon forecast to forecast graph

## Context
- forecast code is in here - https://github.com/openclimatefix/analysis-dashboard/blob/main/src/forecast.py
- need to set DB_URL inorder to run this locally and connect using ssh_tunnel

## Possible Implementation
- Add elexonpy to requirements 
- show 3 different forecast - process type = "Day Ahead", "Intraday Process", "Intraday Total"
- Go to "Forecast page"

### PR DESCRIPTION
**Pull request by 14Richa**
## Description

This pull request makes two updates. First, it adds the `elexonpy` library to the requirements. Second, it adds a feature to show forecasts for three types of processes: Day Ahead, Intraday Process, and Intraday Total. 

PFA the screenshot of the forecast page : 
image

Fixes #172

### CODE REVIEW

**Comment by peterdudfield** [comment]
Good to combine forecasts on one plot.

File: src/forecast.py [hunk 1]
```diff
@@ -213,8 +190,72 @@ def forecast_page():
             )
         )
 
-    st.plotly_chart(fig, theme="streamlit")
-
+    # Fetch and plot Elexon Solar Forecast data
+    st.sidebar.subheader("Select Elexon Forecast Dates")
+    start_datetime_utc = st.sidebar.date_input("Start Date", datetime.utcnow() - timedelta(days=3))
```
Comment (peterdudfield) [inline src/forecast.py:195]: could you use the same start and end times as above, then we dont need these extra options

File: src/forecast.py [hunk 2]
```diff
@@ -213,8 +190,72 @@ def forecast_page():
...
+        line_styles = ["solid", "dash", "dot"]
+        forecasts = [fetch_forecast_data(forecast_generation_wind_and_solar_day_ahead_get, start_datetime_utc, end_datetime_utc, pt) for pt in process_types]
+
+        for i, (forecast, line_style) in enumerate(zip(forecasts, line_styles)):
+            if forecast.empty:
+                st.write(f"No data available for process type: {process_types[i]}")
+                continue
+
+            # Remove NaNs and zero values
+            forecast = forecast[forecast["quantity"].notna() & (forecast["quantity"] > 0)]
```
Comment (peterdudfield) [inline src/forecast.py:210]: could you add a comment, why you are remove nans and zeros?

File: src/forecast.py [hunk 3]
```diff
@@ -213,8 +190,72 @@ def forecast_page():
...
+        return solar_df
+    except Exception as e:
+        st.error(f"Error fetching data for process type '{process_type}': {e}")
+        return pd.DataFrame()
+
+# Initialize Elexon API client
+api_client = ApiClient()
+forecast_api = GenerationForecastApi(api_client)
+forecast_generation_wind_and_solar_day_ahead_get = (
+    forecast_api.forecast_generation_wind_and_solar_day_ahead_get
```
Comment (peterdudfield) [inline src/forecast.py:257]: could you put this initilization wihtin the elexon code block

File: src/forecast.py [hunk 4]
```diff
@@ -171,7 +154,6 @@ def forecast_page():
                     )
                     label = model
 
-                # Make ForecastValue objects with _properties attribute and maybe adjust
```
[the remark this answers is not in the document]
Comment (14Richa) [reply src/forecast.py:174]: Apologies. By mistake. I have added the comments now.

File: src/forecast.py [hunk 5]
```diff
@@ -213,8 +199,100 @@ def forecast_page():
...
+    except Exception as e:
+        st.error(f"Error fetching data for process type '{process_type}': {e}")
+        return pd.DataFrame()
+
+# # Initialize Elexon API client
+# api_client = ApiClient()
+# forecast_api = GenerationForecastApi(api_client)
+# forecast_generation_wind_and_solar_day_ahead_get = (
+#     forecast_api.forecast_generation_wind_and_solar_day_ahead_get
+# )
```
Comment (peterdudfield) [inline src/forecast.py:295]: remove, dont comment out

File: src/forecast.py [hunk 6]
```diff
@@ -213,8 +199,100 @@ def forecast_page():
...
+                else:
+                    # If start_datetime_utc is a date object, convert it to datetime
+                    start_datetime_utc = datetime.combine(start_datetime_utc, datetime.min.time())
+                    end_datetime_utc = max(start_datetime_utc + timedelta(days=7), datetime.utcnow())
+            else:
+                end_datetime_utc = end_datetimes[-1]
+                # Ensure end_datetime_utc is a datetime object
+                if isinstance(end_datetime_utc, datetime.date):
+                    end_datetime_utc = datetime.combine(end_datetime_utc, datetime.min.time())
+
```
Comment (peterdudfield) [inline src/forecast.py:222]: could you refactor this bit to a function, it seems a bit messy in order to get start and end datetimes, but perhaps its the only way

File: src/forecast.py [hunk 7]
```diff
@@ -213,7 +199,108 @@ def forecast_page():
             )
         )
 
-    st.plotly_chart(fig, theme="streamlit")
+        start_datetime_utc, end_datetime_utc = determine_start_and_end_datetimes(start_datetimes, end_datetimes)
+
+        if start_datetime_utc and end_datetime_utc:
+            if start_datetime_utc < end_datetime_utc:
+                # Initialize Elexon API client
```
Comment (peterdudfield) [inline src/forecast.py:206]: do you need these two if statements? Wouldnt they be better dealt with in `determine_start_and_end_datetimes`

File: src/forecast.py [hunk 8]
```diff
@@ -213,8 +199,112 @@ def forecast_page():
...
+    - start_datetime_utc: datetime object in UTC
+    - end_datetime_utc: datetime object in UTC
+    """
+    now = datetime.utcnow()
+
+    # Determine start_datetime_utc
+    if start_datetimes:
+        start_datetime_utc = start_datetimes[0]
+    else:
+        start_datetime_utc = now
```
Comment (peterdudfield) [inline src/forecast.py:287]: should we do `now - 2 days`, so we load something

File: src/forecast.py [hunk 9]
```diff
@@ -213,8 +199,112 @@ def forecast_page():
...
+        end_datetime_utc = end_datetimes[-1]
+    else:
+        end_datetime_utc = start_datetime_utc + timedelta(days=7)
+
+    # Ensure end_datetime_utc is a datetime object
+    if isinstance(end_datetime_utc, date) and not isinstance(end_datetime_utc, datetime):
+        end_datetime_utc = datetime.combine(end_datetime_utc, datetime.min.time())
+
+    # Check if start is before end
+    if start_datetime_utc >= end_datetime_utc:
```
Comment (peterdudfield) [inline src/forecast.py:304]: do we need this? could we do an assert instead? then we dont need the `if` statement above

File: src/forecast.py [hunk 10]
```diff
@@ -213,8 +199,112 @@ def forecast_page():
             )
         )
 
-    st.plotly_chart(fig, theme="streamlit")
-
+        start_datetime_utc, end_datetime_utc = determine_start_and_end_datetimes(start_datetimes, end_datetimes)
+
+        if start_datetime_utc and end_datetime_utc:
+            # Initialize Elexon API client
```
Comment (peterdudfield) [inline src/forecast.py:205]: could we refactor and move to a new file, perhaps in https://github.com/openclimatefix/analysis-dashboard/tree/main/src/plots but call `elexon`, `fetch_forecast_data` could also move there

File: src/forecast.py [hunk 11]
```diff
@@ -213,7 +201,48 @@ def forecast_page():
...
+                forecast = full_time_df.merge(forecast, on='start_time', how='left')
+
+                fig.add_trace(go.Scatter(
+                    x=forecast["start_time"],
+                    y=forecast["quantity"],
+                    mode='lines',
+                    name=f"Elexon {process_types[i]}",
+                    line=dict(color='#318CE7', dash=line_style),
+                    connectgaps=False
+                ))
```
Comment (peterdudfield) [inline src/forecast.py:241]: Could lines 208 to 241, also be moved to elexon_plots. You might have to take in a fig, and return a fig
you can then add a test for this, to check it works

File: src/plots/elexon_plots.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
+import pandas as pd
+from datetime import datetime, date, timedelta
+from plotly import graph_objects as go
+import streamlit as st
+
+def fetch_forecast_data(api_func, start_date, end_date, process_type):
+    try:
```
Comment (peterdudfield) [inline src/plots/elexon_plots.py:7]: Cna you add function descriptions and say what the parameters are.

Could you add type hinting please too,

File: src/plots/elexon_plots.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        # Resample if there's data
+        if not solar_df.empty:
+            solar_df = solar_df.resample("30T")["quantity"].sum().reset_index()
+
+        return solar_df
+    except Exception as e:
+        st.error(f"Error fetching data for process type '{process_type}': {e}")
+        return pd.DataFrame()
+
+def determine_start_and_end_datetimes(start_datetimes, end_datetimes):
```
Comment (peterdudfield) [inline src/plots/elexon_plots.py:31]: could you add type hinting please

File: src/forecast.py [hunk 14]
```diff
@@ -16,7 +16,10 @@
 from plots.utils import (
     get_colour_from_model_name, model_is_probabilistic, get_recent_available_model_names
 )
+from elexonpy.api_client import ApiClient
+from elexonpy.api.generation_forecast_api import GenerationForecastApi
```
Comment (peterdudfield) [inline src/forecast.py:20]: do you need all these imports still?

**Review by peterdudfield — APPROVED** [review]
Looks great, I think just needs some import tidy up and potentially also run blacks with line lenght 100, then I think its good to merge.

Thanks so much for doing this

**Comment by 14Richa** [comment]
> Looks great, I think just needs some import tidy up and potentially also run blacks with line lenght 100, then I think its good to merge.
> 
> Thanks so much for doing this

I've formatted the line length to 100 and tidied up the imports. This PR is ready to merge now. Thanks!

**Review by peterdudfield — COMMENTED** [review]
Last but I think final thing, Could you add one (or more) test for each of the elexon_plot functions. Then we can get it merged in nicely and deployed

File: tests/test_elexon_plot.py [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+    mock_fetch.return_value = pd.DataFrame()
+
+    # Create an empty Plotly figure
+    fig = go.Figure()
+    start_datetime = [datetime(2024, 8, 1)]
+    end_datetime = [datetime(2024, 8, 2)]
+    updated_fig = add_elexon_plot(fig, start_datetime, end_datetime)
+
+    # Assert
+    assert len(updated_fig.data) == 0, "Figure should have no traces added if no data is available"
```
Comment (peterdudfield) [inline tests/test_elexon_plot.py:72]: Greta stuff, could you do one test for this
`fetch_forecast_data` with no mocking
and label it as an integration test like [this](https://github.com/openclimatefix/uk-pv-national-gsp-api/blob/main/src/tests/test_elexon_forecast.py#L59)

File: tests/test_elexon_plot.py [hunk 16]
```diff
@@ -69,4 +71,29 @@ def test_add_elexon_plot_no_data(mock_fetch):
...
+
+    # Call the function with real data
+    result = fetch_forecast_data(forecast_generation_wind_and_solar_day_ahead_get, start_date, end_date, "Day Ahead")
+
+    # Assertions to check the returned DataFrame
+    assert isinstance(result, pd.DataFrame), "Result should be a DataFrame"
+
+    # If data exists for the given dates, the DataFrame shouldn't be empty
+    if not result.empty:
+        assert "start_time" in result.columns, "DataFrame should contain 'start_time' column"
```
Comment (peterdudfield) [inline tests/test_elexon_plot.py:95]: can you just include these asswers, not the empty bit

**Review by peterdudfield — APPROVED** [review]

**Pull request merged**